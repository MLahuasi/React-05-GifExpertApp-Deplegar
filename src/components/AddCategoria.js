import React, { useState } from 'react'
import PropTypes from 'prop-types';

export const AddCategoria = ({ setCategorias }) => {

    const [inputValue, setInputValue] = useState(''); //Si no se envia un parámetro en useState se presenta un error en consola ya que useState es undefined

    const handleInputChange = ( e ) => {
        setInputValue(e.target.value); //Escribe en la caja de texto
    };

    const handleSubmit = (e) => {
        e.preventDefault(); //Previene el refresh de la página

        if( inputValue.trim().length > 2 ){
            setCategorias(cats => [inputValue, ...cats ]);
            setInputValue('');
        }
    }

    return (
        <form onSubmit={handleSubmit}>
            <input
                type="text"
                value={inputValue}
                onChange={handleInputChange}
            />
        </form>
    )
}

AddCategoria.propTypes = {
    setCategorias: PropTypes.func.isRequired
}