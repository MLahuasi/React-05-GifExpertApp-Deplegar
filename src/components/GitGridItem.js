import React from 'react'

export const GitGridItem = ({id, title, url}) => {

    return (
        <div className="card animate__animated animate__bounce">
            <img src={url} alt={ title } />
            <p>{title}</p>
        </div>
    )
}
