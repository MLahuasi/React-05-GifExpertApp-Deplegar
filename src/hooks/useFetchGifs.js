/**
 * Custom Hook
 */

import { useEffect, useState } from 'react';
import { getGifs } from '../helpers/getGifs';

export const useFetchGifs = (categoria) => {

    //Estado de renderización [useState]
    const [state, setState] = useState({
        data: [],
        loading: true
    });


    /**
     * Se ejecuta la renderización solo si existe un cambio
     */
    useEffect(() => {
        getGifs(categoria)
            .then(imgs => {
                console.log(imgs);
                setState({
                    data: imgs,
                    loading: false
                })
            })
    }, [categoria]);


    //el state es la información que se envía en el useState
    return state;
}