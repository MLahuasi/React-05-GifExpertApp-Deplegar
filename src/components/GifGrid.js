import React from 'react';
import { useFetchGifs } from '../hooks/useFetchGifs';
import { GitGridItem } from './GitGridItem';

export const GifGrid = ({ categoria }) => {

    /**
     * Destructuraci√≥n al llamar useFetchGifs
     */
    const { data:images, loading } = useFetchGifs(categoria);
    return (
        <>
            <h3 className='animate__animated animate__fadeIn'>{categoria}</h3>

            {loading && <p className='animate__animated animate__flash'>Loading</p> }

            <div className="card-grid">
                {
                    images.map( img => (
                        <GitGridItem
                            key={img.id}
                            { ...img}
                        />
                    ))
                }
            </div>

        </>
    )
}

