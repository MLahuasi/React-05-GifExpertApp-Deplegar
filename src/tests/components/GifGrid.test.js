import React from 'react';
import '@testing-library/jest-dom';
import { GifGrid } from '../../components/GifGrid';
import { shallow } from 'enzyme';

//Crear mock de useFetchGifs
import { useFetchGifs } from '../../hooks/useFetchGifs';
jest.mock('../../hooks/useFetchGifs');


describe('Test GifGrid', () => {

    const categoria = 'MCU';


    /**
     * Inicializar Test
     */
    beforeEach(() => {
        jest.clearAllMocks();
    });


    test('Debe mostrar el control correctamente ', () => {
        //Invocar al mock
        useFetchGifs.mockReturnValue({
            data: [],
            loading: true
        });


        const wrapper = shallow(<GifGrid categoria={categoria} />);
        expect(wrapper).toMatchSnapshot();
    });

    test('Deben mostrase Items cuando se cragan las imágenes [useFetchGifs]', () => {

        const gifs = [
            {
            id: '1',
            title: 'ABC',
            url: 'https://localhost/imagen1.jpg'
            },
            {
                id: '2',
                title: 'DEF',
                url: 'https://localhost/imagen2.jpg'
            },
            {
                id: '3',
                title: 'GHI',
                url: 'https://localhost/imagen3.jpg'
            }
        ];

        //Invocar al mock
        useFetchGifs.mockReturnValue({
            data: gifs,
            loading: false
        });

        const wrapper = shallow(<GifGrid categoria={categoria} />);
        expect( wrapper ).toMatchSnapshot();
    });

    test('Deben mostrase Items cuando se cragan las imágenes [useFetchGifs] SIN SnapShot', () => {

        const gifs = [
            {
                id: '4',
                title: 'JKL',
                url: 'https://localhost/imagen4.jpg'
            },
            {
                id: '5',
                title: 'MNO',
                url: 'https://localhost/imagen5.jpg'
            },
            {
                id: '6',
                title: 'PQR',
                url: 'https://localhost/imagen6.jpg'
            }
        ];

        //Invocar al mock
        useFetchGifs.mockReturnValue({
            data: gifs,
            loading: false
        });

        const wrapper = shallow(<GifGrid categoria={categoria} />);

        expect(wrapper).toMatchSnapshot();
        expect( wrapper.find('p').exists() ).toBe(false);
        //Existen 3 elementos de tipo GitGridItem
        expect(wrapper.find('GitGridItem').length).toBe( gifs.length );
    });

})
