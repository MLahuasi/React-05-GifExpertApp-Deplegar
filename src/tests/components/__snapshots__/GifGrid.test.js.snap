// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Test GifGrid Debe mostrar el control correctamente  1`] = `
<Fragment>
  <h3
    className="animate__animated animate__fadeIn"
  >
    MCU
  </h3>
  <p
    className="animate__animated animate__flash"
  >
    Loading
  </p>
  <div
    className="card-grid"
  />
</Fragment>
`;

exports[`Test GifGrid Deben mostrase Items cuando se cragan las im√°genes [useFetchGifs] 1`] = `
<Fragment>
  <h3
    className="animate__animated animate__fadeIn"
  >
    MCU
  </h3>
  <div
    className="card-grid"
  >
    <GitGridItem
      id="1"
      key="1"
      title="ABC"
      url="https://localhost/imagen1.jpg"
    />
    <GitGridItem
      id="2"
      key="2"
      title="DEF"
      url="https://localhost/imagen2.jpg"
    />
    <GitGridItem
      id="3"
      key="3"
      title="GHI"
      url="https://localhost/imagen3.jpg"
    />
  </div>
</Fragment>
`;
